crates/iceberg/src/io.rs

```diff
@@ -268,6 +268,16 @@ impl OutputFile {
             .await?)
     }
 
+    /// Delete file.
+    pub async fn delete(&self) -> Result<()> {
```

Comment:
This seems unsound to me. If it's deleted, we should consume it, so its argument should be `self` rather `&self`

Reply:
I'm thinking of the bad cases: If the `delete` returns error, how can we retry this operation?

Reply:
> I'm thinking of the bad cases: If the delete returns error, how can we retry this operation?

This delete operation used in the cases:
The writer created and it didn't write anything, so we deleted it to avoid the empty file. If delete fails here, can we just skip this rare case?🤔

But I find that there maybe a wrong hypothesis. Does [writer in opendal](https://docs.rs/opendal/latest/opendal/struct.Operator.html?search=writer#method.writer) create the file?

Reply:
It seems more reasonable to move it to `FileIO`. An `OutputFile` may not exists yet.

Reply:
> Does [writer in opendal](https://docs.rs/opendal/latest/opendal/struct.Operator.html?search=writer#method.writer) create the file?

It depends. AFAIK, opendal on s3 won't create file unless `close` has returned `Ok()`.

By the way, the use case you mentioned here more like [`Writer::abort`](https://opendal.apache.org/docs/rust/opendal/struct.Writer.html#method.abort).

Reply:
> By the way, the use case you mentioned here more like [Writer::abort](https://opendal.apache.org/docs/rust/opendal/struct.Writer.html#method.abort).

Nice. We replace delete with `abort`.

Reply:
We will pass the writer to parquet writer so we can't use `abort` directly.🥵 And use the FileIO to delete the file now.

Reply:
I find that we don't need to delete empty files in this writer. We can do this thing in the upper writer and make this writer be simple. This writer only responded to get an output file and write it. 

And I find that for FileWriterBuilder should accept the output file when built. We hope to clone the builder and rebuild a writer using a new output file, which means that the builder state should be bound with any output file.